NotificationService/NotificationService.BusinessLibrary/Providers/MSGraphProvider.cs

```diff
@@ -124,43 +124,27 @@ public async Task<IList<NotificationBatchItemResponse>> ProcessEmailRequestBatch
         }
 
         /// <inheritdoc/>
-        public async Task<bool> SendEmailNotification(AuthenticationHeaderValue authenticationHeaderValue, EmailMessagePayload payLoad, string notificationId)
+        public async Task<ResponseData<string>> SendEmailNotification(AuthenticationHeaderValue authenticationHeaderValue, EmailMessagePayload payLoad, string notificationId)
         {
             this.logger.TraceInformation($"Started {nameof(this.SendEmailNotification)} method of {nameof(MSGraphProvider)}.");
             this.httpClient.DefaultRequestHeaders.Authorization = authenticationHeaderValue;
             var requestPayLoad = JsonConvert.SerializeObject(payLoad, this.jsonSerializerSettings);
             HttpResponseMessage response = null;
-            bool isSuccess = false;
             response = await this.httpClient.PostAsync(
                     $"{this.mSGraphSetting.BaseUrl}/{this.mSGraphSetting.GraphAPIVersion}/{this.mSGraphSetting.SendMailUrl}",
                     new StringContent(requestPayLoad, Encoding.UTF8, ApplicationConstants.JsonMIMEType)).ConfigureAwait(false);
 
-            this.logger.TraceInformation($"Method {nameof(this.SendEmailNotification)}: Completed Graph Send Email Call.");
-            var responseHeaders = response.Headers.ToString();
+            this.logger.TraceInformation($"Method {nameof(this.SendEmailNotification)}: Completed Graph Send Email Call for notificationId : {notificationId}");
 
-            if (response.IsSuccessStatusCode)
-            {
-                // Read and deserialize response.
-                var content = await response.Content.ReadAsStringAsync().ConfigureAwait(false);
-                isSuccess = true;
-            }
-            else if (response.StatusCode == HttpStatusCode.TooManyRequests || response.StatusCode == HttpStatusCode.RequestTimeout)
-            {
-                isSuccess = false;
-            }
-            else
-            {
-                string content = string.Empty;
-                if (response != null)
-                {
-                    content = await response.Content.ReadAsStringAsync().ConfigureAwait(false);
-                }
+            var responseData = await GetResponseData(response).ConfigureAwait(false);
 
-                throw new System.Exception($"An error occurred while sending notification id: {notificationId}. Details: {content}");
+            if (responseData == null || (!responseData.Status && !(responseData.StatusCode == HttpStatusCode.TooManyRequests || responseData.StatusCode == HttpStatusCode.RequestTimeout)))
+            {
+                throw new System.Exception($"An error occurred while sending notification id: {notificationId}. Details: {responseData?.Result}");
```

Comment:
you are throwing the exception here, how come this status will be returned to caller, where you are checking this too many requests check in caller? 

Test it completly once. I doubt this will work

Reply:
If you see the code it is same as the delete one, only i am doing additional is i am returning the ResponseData object instead of bool value, as the responsedata was required to capture statuscode. The exception was already handled in ProcessEntitiesIndividually (caller). if you see the line 159 (deleted) this was also thrown based on same condition as in new one.